col*: support json type in vectorized engine using datum

Azhng opened the pull request:
Previously, when vectorized engine encounters an unsupported data type it falls
back to row execution engine. We introduce new DatumVec in order to use
tree.Datum to support all data types. This PR only introduces support for JSON
types and overloads for datum comparison.

Since we do not want to introduce dependency from `coldata` to `tree` package,
the implementation of DatumVec actually lives in `colmem` package to implement
an interface declared in `coldata` package.


Release note (sql change): vectorized engine now supports JSON columns.

Addresses  #45612.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/47536)
<!-- Reviewable:end -->

Azhng commented:
This PR does not address the issue where `pkg/col/coldata` is now depending on `pkg/sql`. 

As discussed in offline, we might need to extract datum logic and encoding logic out from `pkg/sql`

yuzefovich reviewed (commented):
> Reviewed 1 of 39 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47536)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng)

___
*[pkg/col/coldata/batch.go, line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47536#-M4zTuRzDyucWjczSSXY:-M4zTuRzDyucWjczSSXZ:bbhgp52) ([raw file](https://github.com/cockroachdb/cockroach/blob/57800c0ff6ec2c47cfb88e2fe45c161ed31dac3a/pkg/col/coldata/batch.go#L115)):*
> ```Go
> // size and evalCtx. Use for operators that have a precisely-sized output batch
> // and also responsible for instantiating the first memColumn of DatumVec type.
> func NewMemBatchWithSizeAndEvalCtx(types []coltypes.T, size int, evalCtx *tree.EvalContext) Batch {
> ```

Dependency of `coldata` package on `tree.EvalContext` seems bad. I think we should try to figure out a way to avoid it. One idea is that we introduce `DatumVec` as an interface in `coldata` package while actual implementation would live in `colexec` package, something like:
```
package coldata

type DatumVec interface {
  func CompareDatum(other tree.Datum) int
  // other methods
}
```
Then in `vec.go` we would have:
```
type Vec interface {
  ...
  Datum() DatumVec
}

func (m *memColumn) Datum() DatumVec {
	return m.col.(DatumVec)
}
```
I'm worried that extracting `tree.EvalContext` out of `tree` package will not be that easy.
___
*[pkg/col/colserde/file.go, line 386 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47536#-M4zXLxE8Nh-dSy1gOru:-M4zXLxFDuU7Nkenut0w:ba9sai5) ([raw file](https://github.com/cockroachdb/cockroach/blob/57800c0ff6ec2c47cfb88e2fe45c161ed31dac3a/pkg/col/colserde/file.go#L386)):*
> ```Go
> 			fbTypOffset = arrowserde.BinaryEnd(fb)
> 			fbTyp = arrowserde.TypeInterval
> 		case coltypes.Datum:
> ```

I feel like this case should be exactly the same as for `coltypes.Bytes`.


<!-- Sent from Reviewable.io -->

yuzefovich reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/47536)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @Azhng)

___
*[pkg/col/coldata/batch.go, line 115 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/47536#-M4zTuRzDyucWjczSSXY:-M4zZVoQ5VEzrPhrGSHV:b-ufn4yj) ([raw file](https://github.com/cockroachdb/cockroach/blob/57800c0ff6ec2c47cfb88e2fe45c161ed31dac3a/pkg/col/coldata/batch.go#L115)):*
<details><summary><i>Previously, yuzefovich wrote…</i></summary><blockquote>

Dependency of `coldata` package on `tree.EvalContext` seems bad. I think we should try to figure out a way to avoid it. One idea is that we introduce `DatumVec` as an interface in `coldata` package while actual implementation would live in `colexec` package, something like:
```
package coldata

type DatumVec interface {
  func CompareDatum(other tree.Datum) int
  // other methods
}
```
Then in `vec.go` we would have:
```
type Vec interface {
  ...
  Datum() DatumVec
}

func (m *memColumn) Datum() DatumVec {
	return m.col.(DatumVec)
}
```
I'm worried that extracting `tree.EvalContext` out of `tree` package will not be that easy.
</blockquote></details>

Well, maybe it's not that easy to separate `tree.Datum` from `tree.EvalContext`. It seems like we need to think hard of what the package structure we want and what should the dependencies be.


<!-- Sent from Reviewable.io -->

Azhng committed:
coldata, colflow, colexec, colserde: support JSON type for vectorized engine
using tree.Datum

Previously, when vectorized engine encounters an unsupported data type it falls
back to row execution engine. We introduce new DatumVec in order to use
tree.Datum to support all data types. This PR only introduces support for JSON
types and overloads for datum comparison.

Since we do not want to introduce dependency from `coldata` to `tree` package,
the implementation of DatumVec actually lives in `colmem` package to implement
an interface declared in `coldata` package.

Release note (sql change): vectorized engine now supports JSON columns.

Azhng commented:
Here is a summary of the current progress my my attempt to remove `coldata` dependency on `tree` :
* Introduced `ColumnFactory` interface which will be responsible to returns a `column` given a `coltype`. There are two implementation of `ColumnFactory`:
  * `StandardColumnFactory`, which is implemented within `coldata` package that construct all current coltypes.
  * `ExtendedColumnFactory`, implemented within `colmem` package (possibly this can be moved to `colexec`) that constructs all current coltypes and the new `DatumVec` which represents the new datum vector. `ExtendedColumnFactory` will store `evalCtx` as it's parameter and it will be used to construct `DatumVec`. 
* As a result `column` is renamed to `Column` in order to be exported 
* Also as the result, now `NewMemBatch` interface is modified to accept the `ColumnFactory` interface.
* Allocator interface is also changed to accept the `ColumnFactory`. This allows `allocator`'s `NewMemBatch` interface to remain the same.
* Introduced new `DatumContainer` interface which lives in `coldata` package, it defines the interface which can be consumed by `Vec` and `execgen`. Unfortunately in order to avoid importing `tree`, I would have to rely on duck-typing and runtime type cast using `interface{}` some places, (`DatumContainerElem` in particular)
* The implementation of `DatumContainer` is `DatumVec` which lives in `colmem` package along with its unit test. 

Looks like current CI build time is longer, this could potentially be due to some `*_test` packages importing the `colmem` which in terms importing `tree` package. This can potentially be improved to use `coldata.StandardColumnFactory` if the test does not require to use JSON type.

Current CI failure is due to `workload` importing `colmem`, and `coltypes` importing `json`, which are not whitelisted. 
* The proposed method of preventing `coldata` importing `tree` cannot prevent `workload` importing `colmem`. This can be partially address to use `coldata.StandardColumnFactory`, however, this will not allow the `workload` to generate json columns.
* `coltypes` importing `json` is difficult to resolve because of the use of type switch in `FromGoType`. This can be potentially solved using `reflect` inside the `default` clause, though this is probably too slow unless it's part of the testing code.

yuzefovich commented:
Superseded by #48422
